File: Day1/Go/dial.go
package main

// Dial represents a circular dial with a specific size and position
type dial struct {
	Size             int
	Position         int
	TimesReachedZero int
	StepValue        int
}

// NewDial creates a new Dial instance with the given size and starting position
func NewDial(size int, startPosition int) *dial {
	return &dial{
		Size:     size,
		Position: startPosition,
	}
}

func (d *dial) Turn(steps int, direction string) {
	// Determine step value based on direction
	d.StepValueCalc(direction)

	// Update position with wrap-around logic
	d.Position += steps * d.StepValue
	d.Position = ((d.Position + d.Size) + d.Size) % d.Size

	// Check if position is zero
	d.CheckZero()
}

func (d *dial) TurnCountAllZeros(steps int, direction string) {
	// Determine step value based on direction
	d.StepValueCalc(direction)

	// Update position step by step to count all zeros
	for i := 0; i < steps; i++ {
		d.Position += d.StepValue
		d.Position = ((d.Position + d.Size) + d.Size) % d.Size
		d.CheckZero()
	}
}

func (d *dial) StepValueCalc(direction string) {
	if direction == "R" {
		d.StepValue = 1
	} else if direction == "L" {
		d.StepValue = -1
	}
}

func (d *dial) CheckZero() {
	if d.Position == 0 {
		d.TimesReachedZero++
	}
}

File: Day1/CSharp/dial.cs
namespace dial;

public class Dial
{
    public int Size { get; }
    public int Position { get; private set; }
    public int TimesZeroReached { get; set; }

    public Dial(int size, int startPosition)
    {
      Size = size;
      Position = startPosition;
      TimesZeroReached = 0;
    }

    public void Turn(int steps, string direction)
    {
      int stepValue = (direction == "R") ? 1 : -1;
      Position += steps * stepValue;
      Position = ((Position % Size) + Size) % Size;
      if (Position == 0)
      {
        TimesZeroReached++;
      }
    }

    public void TurnCountAllZero(int steps, string direction)
    {
      int stepValue = (direction == "R") ? 1 : -1;
      for (int i = 0; i < steps; i++)
      {
        Position += stepValue;
        Position = ((Position % Size) + Size) % Size;
        if (Position == 0)
        {
          TimesZeroReached++;
        }
      }
    }
}